File: 09.Greed贪心/leetcode_3635_earliestFinishTime.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

class Solution {
public:
    int earliestFinishTime(vector<int>& landStartTime, vector<int>& landDuration, vector<int>& waterStartTime, vector<int>& waterDuration) {
        int l = landStartTime.size(), w = waterStartTime.size();
        int minLand = INT_MAX, minWater = INT_MAX;
        for (int i=0; i<l; i++){
            minLand = min(minLand, landStartTime[i]+landDuration[i]);
        }
        for (int i=0; i<w; i++){
            minWater = min(minWater, waterStartTime[i]+waterDuration[i]);
        }
        int res = INT_MAX;
        for (int i=0; i<w; i++){
            res = min(res, max(minLand, waterStartTime[i])+waterDuration[i]);
        }
        for (int i=0; i<l; i++){
            res = min(res, max(minWater, landStartTime[i])+landDuration[i]);
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<int> landStartTime = {2,8}, landDuration = {4,1}, waterStartTime = {6}, waterDuration = {3};
    cout << s.earliestFinishTime(landStartTime, landDuration, waterStartTime, waterDuration) << endl;
}

File: 09.Greed贪心/leetcode_3635_earliestFinishTime.py
from typing import List

class Solution:
    def earliestFinishTime(self, landST: List[int], landD: List[int], waterST: List[int], waterD: List[int]) -> int:
        l, w = len(landST), len(waterST)
        minLand, minWater = 10**18, 10**18
        for i in range(l):
            minLand = min(minLand, landST[i]+landD[i])
        for i in range(w):
            minWater = min(minWater, waterST[i]+waterD[i])
        
        res = 10**18
        for i in range(w):
            res = min(res, max(minLand, waterST[i])+waterD[i])
        for i in range(l):
            res = min(res, max(minWater, landST[i])+landD[i])
        return res

if __name__ == "__main__":
    s = Solution()
    landST, landD = [2,8], [4,1]
    waterST, waterD = [6], [3]
    print(s.earliestFinishTime(landST, landD, waterST, waterD))
